File: Legacy/Leetcode/94BinaryTreeInorderTraversal.py
# Given a binary tree, return the inorder traversal of its nodes' values.
#
# For example:
# Given binary tree [1,null,2,3],
#    1
#     \
#      2
#     /
#    3
# return [1,3,2].

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root == None:
            return []
        stack = []
        ans = []
        visit = {}
        stack.append(root)
        while stack:
            node = stack[-1].left
            while node != None and node not in visit:
                stack.append(node)
                node = node.left
            node = stack.pop()
            ans.append(node.val)
            visit[node] = True
            if node.right != None:
                stack.append(node.right)
        return ans

File: C++/Tree/94. Binary Tree Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> stk;
        while (root || !stk.empty()) {
            while (root) {
                stk.push(root);
                root = root->left;
            }
            res.push_back(stk.top()->val);
            root = stk.top()->right;
            stk.pop();
        }
        return res;
    }
};

// 递归

// class Solution {
// public:
//     vector<int> inorderTraversal(TreeNode* root) {
//         vector<int> res;
//         if (!root) return res;
//         res = inorderTraversal(root->left);
//         res.push_back(root->val);
//         vector<int> right = inorderTraversal(root->right);
//         res.insert(res.end(), right.begin(), right.end());
//         return res;
//     }
// };